src/build/cargo.rs

```diff
@@ -56,98 +62,132 @@ impl Internals {
+        trace!("Cargo compilation options:\n{:?}", opts);
+        let rustflags = prepare_cargo_rustflags(&rls_config);
+
+        // Warn about invalid specified bin target or package depending on current mode
+        // TODO: Return client notifications along with diagnostics to inform the user
+        if !rls_config.workspace_mode {
+            let cur_pkg_targets = ws.current().unwrap().targets();
+
+            if let Some(ref build_bin) = rls_config.build_bin {
+                let mut bins = cur_pkg_targets.iter().filter(|x| x.is_bin());
+                if let None = bins.find(|x| x.name() == build_bin) {
+                    warn!("cargo - couldn't find binary `{}` specified in `build_bin` configuration", build_bin);
+                }
+            }
+        } else {
+            for package in &opts.package {
+                if let None =  ws.members().find(|x| x.name() == package) {
+                    warn!("cargo - couldn't find member package `{}` specified in `analyze_package` configuration", package);
                 }
-            } else {
-                vec![]
             }
-        };
-
-        let mut opts = CompileOptions::default(&config, CompileMode::Check);
-        if rls_config.build_lib {
-            opts.filter = CompileFilter::new(true, &[], false, &[], false, &[], false, &[], false); 
-        } else if !bins.is_empty() {
-            opts.filter = CompileFilter::new(false, &bins, false, &[], false, &[], false, &[], false);
-        }
-        if let Some(ref target) = rls_config.target {
-            target_string = target.clone();
-            opts.target = Some(&target_string);
         }
-        opts
+
+        (opts, rustflags)
+    };
+
+    let spec = Packages::from_flags(opts.all, &opts.exclude, &opts.package)
+        .expect("Couldn't create Packages for Cargo");
+
+    let compile_opts = CompileOptions {
+        target: opts.target.as_ref().map(|t| &t[..]),
+        spec: spec,
+        filter: CompileFilter::new(opts.lib,
+                                &opts.bin, opts.bins,
+                                // TODO: Support more crate target types
+                                &[], false, &[], false, &[], false),
+        .. CompileOptions::default(&config, CompileMode::Check)
     };
-    compile_with_exec(&ws, &opts, Arc::new(exec)).expect("could not run cargo");
+
+    env::set_var("RUSTFLAGS", rustflags);
+    compile_with_exec(&ws, &compile_opts, Arc::new(exec)).expect("could not run cargo");
 }
 
 struct RlsExecutor {
     compilation_cx: Arc<Mutex<CompilationContext>>,
     cur_package_id: Mutex<Option<PackageId>>,
     config: Arc<Mutex<Config>>,
+    workspace_mode: bool,
+    /// Packages which are directly a member of the workspace, for which
+    /// analysis and diagnostics will be provided
+    member_packages: Mutex<HashSet<PackageId>>,
+    /// JSON compiler messages emitted for each primary compiled crate
+    compiler_messages: Arc<Mutex<Vec<String>>>,
```

Comment:
Given these fields are only used in workspace mode, we should use an enum for the mode rather than a bool and have these as fields on an enum variant. (We can do this later, doesn't need to block merging now since these fields will probably change as you work on this feature).